.github/vmctl

```diff
@@ -0,0 +1,131 @@
+  lxc launch "images:ubuntu/${VM_SERIES}/desktop" "${VM_NAME}" --vm \
+    -c limits.cpu="${VM_CPUS}" -c limits.memory="${VM_MEM_GIB}GiB" -d root,size="${VM_DISK_GIB}GiB"
+
+  # Wait for the VM agent to be running
+  while ! lxc exec test-vm -- cat /etc/hostname &>/dev/null; do 
+    sleep 2
+  done
+
+  # Push script runner into the VM
+  lxc file push "$(build_runner_script)" test-vm/bin/vmctl-runner
+  lxc exec test-vm -- chmod 755 /bin/vmctl-runner
+
+  # Install gnome-screenshot
+  lxc exec test-vm -- apt-get update
+  lxc exec test-vm -- apt-get install -y gnome-screenshot
+
+  # Kill the gnome initial setup wizard
+  pid="$(lxc exec test-vm -- pidof gnome-initial-setup)"
+  lxc exec test-vm -- kill -9 "$pid"
+}
+
+# Pushes a local snap file into the VM and installs it.
+push_snap() {
+  local filename="$1"
+  lxc file push "${filename}" "test-vm/home/ubuntu/${filename}"
+  lxc exec test-vm -- snap install --dangerous "/home/ubuntu/${filename}"
+}
+
+# Exec a command in the VM using the wrapper.
+exec_in_vm() {
+  lxc exec test-vm -- sudo -u ubuntu bash -c "/bin/vmctl-runner $@"
+}
+
+# Uploads the specified file to imgur and prints the URL to the image
+upload_screenshot() {
+  curl -s -X POST -F "image=@${1}" https://api.imgur.com/3/upload | jq -r '.data.link'
+}
+
+# Takes a screenshot of the full screen and pulls the file back from the VM to a file
+# named "screenshot-screen.png", and uploads to imgur, returning the URL
+screenshot_full() {
+  # Take a screenshot of the whole screen in the VM
+  exec_in_vm "gnome-screenshot -f /home/ubuntu/screen.png"
+  # Pull the screenshot back to the host
+  lxc file pull test-vm/home/ubuntu/screen.png screenshot-screen.png
+  # Upload the screenshot to imgur and output the URL
+  upload_screenshot screenshot-screen.png
+}
+
+# Takes a screenshot of the active window and pulls the file back from the VM to a file
+# named "screenshot-window.png", and uploads to imgur, returning the URL
+screenshot_window() {
+  # Take a screenshot of the active window in the VM
+  exec_in_vm "gnome-screenshot -w -f /home/ubuntu/window.png"
+  # Pull the screenshot back to the host
+  lxc file pull test-vm/home/ubuntu/window.png screenshot-window.png
+  # Upload the screenshot to imgur and output the URL
+  upload_screenshot screenshot-window.png
+}
+
+# Print the usage statement and exit the program
+usage() {
+  echo "Usage: vmctl [prepare | push-snap <file> | exec <string> | screenshot-full | screenshot-window"]
+  exit 1
+}
+
+# Parse the subcommand and exit if empty, printing the usage
+command="${1:-}"; shift
+[[ -z "$command" ]] && usage
+
+case "$command" in
+  "prepare")
+    enable_kvm_gha
+    prepare_vm
+    ;;
+  "push-snap")
+    push_snap "$1"
+    ;;
+  "exec")
+    exec_in_vm "$@"
+    ;;
+  "screenshot-full")
+    screenshot_full
+    ;;
+  "screenshot-window")
+    screenshot_window
+    ;;
+  *)
+    echo "Usage: vmctl prepare|push-snap <file>|exec <string>|screenshot-full|screenshot-window"
```

Comment:
Nitpick: put this in a variable instead of duplicating it.

Reply:
Oops - I'd even created the function to repeat it... :man_facepalming: